modules/gapi/test/internal/gapi_int_pattern_matching_test.cpp

```diff
@@ -994,4 +998,794 @@ TEST(PatternMatching, adeSmokeTest)
+
+struct PatternMatchingIntegrationEndlessLoops :
+    TestWithParam<std::tuple<CompCreator, cv::GCompileArgs>> {};
+TEST_P(PatternMatchingIntegrationEndlessLoops, TestTransformationThrows) {
+    cv::Size in_sz(640, 480);
+    cv::Mat input(in_sz, CV_8UC3);
+    cv::randu(input, cv::Scalar::all(0), cv::Scalar::all(100));
+    cv::Mat orig_output, transformed_output;
+
+    auto mainC = std::get<0>(GetParam())();  // get new copy with new Priv
+    auto args = std::get<1>(GetParam());
+    cv::gimpl::GCompiler compiler(mainC, cv::descr_of(cv::gin(input)), std::move(args));
+
+    EXPECT_THROW(compiler.compile(), std::exception);  //  should throw here
+}
+
+GAPI_TRANSFORM(EndlessLoopTransform1, <cv::GMat(cv::GMat)>, "pattern in substitute")
+{
+    static cv::GMat pattern(const cv::GMat& in)
+    {
+        return cv::gapi::resize(in, cv::Size(100, 100), 0, 0, cv::INTER_LINEAR);
+    }
+
+    static cv::GMat substitute(const cv::GMat& in)
+    {
+        cv::GMat b, g, r;
+        std::tie(b, g, r) = cv::gapi::split3(in);
+        auto resize = std::bind(&cv::gapi::resize,
+            std::placeholders::_1, cv::Size(100, 100), 0, 0, cv::INTER_LINEAR);
+        cv::GMat out = cv::gapi::merge3(resize(b), resize(g), resize(r));
+        return out;
+    }
+};
+GAPI_TRANSFORM(EndlessLoopTransformChain1, <cv::GMat(cv::GMat)>, "Resize -> Custom Resize")
+{
+    static cv::GMat pattern(const cv::GMat& in)
+    {
+        return cv::gapi::resize(in, cv::Size(100, 100));
+    }
+
+    static cv::GMat substitute(const cv::GMat& in)
+    {
+        return MyInterleavedResize::on(in, cv::Size(100, 100), cv::INTER_AREA);
+    }
+};
+GAPI_TRANSFORM(EndlessLoopTransformChain2, <cv::GMat(cv::GMat)>, "Custom Resize -> Resize")
+{
+    static cv::GMat pattern(const cv::GMat& in)
+    {
+        return MyInterleavedResize::on(in, cv::Size(100, 100), cv::INTER_AREA);
+    }
+
+    static cv::GMat substitute(const cv::GMat& in)
+    {
+        return cv::gapi::resize(in, cv::Size(100, 100));
+    }
+};
+
+// FIXME: add more tests for different kinds of loops (when the logic is there)
+INSTANTIATE_TEST_CASE_P(All, PatternMatchingIntegrationEndlessLoops,
+    Values(
+        std::make_tuple(
+            [] () {
+                GMat in;
+                GMat tmp = cv::gapi::resize(in, cv::Size(100, 100), 0, 0, cv::INTER_LINEAR);
+                GMat out = cv::gapi::bitwise_not(tmp);
+                return cv::GComputation(cv::GIn(in), cv::GOut(out));
+            },
+            cv::compile_args(cv::gapi::kernels<EndlessLoopTransform1>())
+        ),
+        std::make_tuple(
+            [] () {
+                GMat in;
+                GMat out = cv::gapi::resize(in, cv::Size(100, 100));
+                return cv::GComputation(cv::GIn(in), cv::GOut(out));
+            },
+            cv::compile_args(
+                cv::gapi::kernels<EndlessLoopTransformChain2, EndlessLoopTransformChain1>())
+        )
+    ));
```

Comment:
Test cases are nice, but please make it more "plain" (more cohesion, less coupling)